ui/lesson/OpCodeParser/index.ts

```diff
@@ -0,0 +1,440 @@
+/* eslint-env browser */
+'use strict'
+
+// Types
+export type InterpreterRef = React.MutableRefObject<HTMLInputElement>
+export type InterpreterDivRef = React.MutableRefObject<HTMLDivElement>
+export type InterpreterAreaRef = React.MutableRefObject<HTMLTextAreaElement>
+export type InterpreterState = {
+  exec_ptr: number
+  stack: any[]
+  negate: number
+  state: boolean[]
+  script: string[]
+  finished?: boolean
+}
+
+interface Refs {
+  scriptFieldRef: InterpreterRef
+  initialStackRef: InterpreterRef
+  executionStackRef: InterpreterAreaRef
+  currentOpCodeRef: InterpreterDivRef
+  heightRef: InterpreterRef
+}
+
+const initialState: InterpreterState = {
+  exec_ptr: -1,
+  stack: [],
+  negate: 0,
+  state: [],
+  script: [],
+}
+
+// UI Update Functions
+const updateStack = (executionStack: InterpreterAreaRef, stack: any[]) => {
+  if (executionStack.current) {
+    executionStack.current.value = stack
+      .map((a) => a)
+      .reverse()
+      .join('\n')
+  } else {
+    console.error('executionStack.current is null')
+  }
+}
+
+const currentOp = (currentOpcode: InterpreterDivRef, opcode: string) => {
+  if (currentOpcode.current) {
+    currentOpcode.current.innerHTML = `<span style="color:blue">${opcode}</span>`
+  }
+}
+
+const error = (currentOpcode: InterpreterDivRef, message: string) => {
+  currentOpcode.current.innerHTML = `<span style="color:red">ERROR: ${message}</span>`
+}
+
+const highlight = (
+  scriptField: InterpreterRef,
+  script: string[],
+  exec_ptr: number
+) => {
+  if (exec_ptr < 0) {
+    if (scriptField.current) {
+      scriptField.current.value = script.join(' ')
+    }
+    return
+  }
+
+  const opcode = script[exec_ptr]
+  const highlight = '->[' + opcode + ']<-'
+  const copy = Array.from(script)
+  copy[exec_ptr] = highlight
+  scriptField.current.value = copy.join(' ')
+}
+
+// Interpreter Functions
+const getKey = (keyData: string): string => {
+  if (!keyData) {
+    throw new Error('Missing key')
+  }
+  const keyMatch = /PUBKEY\((.*?)\)/.exec(keyData)
+  if (!keyMatch || keyMatch.length !== 2) {
+    throw new Error(`Invalid public key: ${keyData}`)
+  }
+  return keyMatch[1]
+}
+
+const getSig = (sigData: string): string => {
+  if (!sigData) {
+    throw new Error('Missing sig')
+  }
+  const sigMatch = /SIG\((.*?)\)/.exec(sigData)
+  if (!sigMatch || sigMatch.length !== 2) {
+    throw new Error(`Invalid signature: ${sigData}`)
+  }
+  return sigMatch[1]
+}
+
+const verify = (stack: any[], error: (msg: string) => void): boolean => {
+  if (stack.length !== 1) {
+    error('Script evaluates false: stack length is not exactly 1')
+    return false
+  }
+  if (stack[0] === false || parseInt(stack[0]) === 0) {
+    error('Script evaluates false: top stack item is false')
+    return false
+  }
+
+  return true
+}
+
+// Opcode Functions
+const opFunctions: { [key: string]: Function } = {
```

Comment:
One thing we might want to consider is to separate these into operations and objects so that we can more easily make sense of these when trying to create this styling element for the stack 

![image](https://github.com/saving-satoshi/saving-satoshi/assets/108441023/8de5102b-79f9-44d5-8fad-908245b1a323)

alternatively we might want to create them as objects to allow us to add some key like `isOperation`

Reply:
OP_1 and OP_2 simply add to the stack while OP_ADD modifies it

Reply:
Oh yh, and operation clears out on the next stack

Reply:
I think we could use the headings from this page https://en.bitcoin.it/wiki/Script to organize them, it would help with thinking of how to style / color them

Reply:
there are
- constants
- flow control
- stack
- splice
- bitwise logic
- arithmetic
- crypto
- locktime
- reserved words

Reply:
ultimately I think this logic should be taken out and put separately and then imported here

Reply:
Should we do this after the stack has been updated
for example

`export const opCodeTypes = (opcode: string) => {
  if (opcode.match(/OP_[0-9]*/)) {
    return { opcode, operation: 'constants' }
  }
  switch (opcode) {
    case 'OP_DUP':
      return { opcode, operation: 'stack' }
    case 'OP_ADD':
      return { opcode, operation: 'arithmetic' }
    case 'OP_EQUAL':
      return { opcode, operation: 'bitwise' }
      }
      `
      
      I have something like this, isn't it best to check on the stack and not during the operation